nixos/hyprland: Add systemd.setPath.enable option to include system and user bin directory in PATH

JohnRTitor opened the pull request:
## Description of changes
This PR adds systemd.setPath.enable option to hyprland module which sets systemd default environment path to include the current system's bin, profile bin and current user's bin.
This is needed in Hyprland only setups, where opening links in applications like VS code do not work without explicitly importing PATH using systemctl.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

JohnRTitor committed:
nixos/hyprland: add systemd.setPath.enable option

This commit adds systemd.setPath.enable option to hyprland module
which sets the systemd path to include the current system's bin
This is needed in Hyprland only setups, where opening links in applications like VS code do not work.

JohnRTitor commented:
@wozeparrot @fufexan Hyprland package maintainers

fufexan reviewed (approved):
LGTM.

One thing though, `mdDoc` is now the default and can be omitted. Can you perhaps also remove any occurrence of it in a separate commit?

JohnRTitor committed:
nixos/hyprland: removed references to mdDoc since it is the default

JohnRTitor committed:
nixos/hyprland: Add possible bin directory paths, ie, user packages bin

If user chooses, they can `exec-once=dbus-update-activation-environment --systemd --all` in hyprland.conf
To import all path variables from the system's environment to systemd's environment

Also set option example to false

JohnRTitor commented:
Done. Also added possible binary dirs such as the user packages bin, profiles/default/bin...

@fufexan please check.

fufexan reviewed (commented):
The commits should start with `nixos/hyprland: ...` and possibly have more concise titles. Further explanations can go into the commit message.

JohnRTitor commented on nixos/modules/programs/wayland/hyprland.nix:
```diff
@@ -63,6 +72,12 @@ in
+
+    systemd = mkIf cfg.systemd.setPath.enable {
+      user.extraConfig = ''
+        DefaultEnvironment="PATH=$PATH:/run/current-system/sw/bin:/run/wrappers/bin:/nix/profile/bin:/etc/profiles/per-user/$USER/bin:/nix/var/nix/profiles/default/bin"
```
Thanks for the suggestion, just one thing though, /etc/profiles/per-user/$USER/bin is used for storing user specific packages and also home-manager packages (in some cases).

Running `ls /etc/profiles/per-user/masum/bin/` on my system shows up the home manager and user packages.
```
accessdb        deluge-web  git-credential-netrc  lexgrog      nslookup    sar                   soffice               which
alacritty       delv        git-cvsserver         libreoffice  nsupdate    sbase                 starship              zsh
ani-cli         difft       git-http-backend      lsof         pcilmr      scalar                swriter               zstd
```

And you are right, I should have been more careful with commit naming. :D

fufexan replied:
Oh, you're right about `/etc/profiles`. You can leave that in then.

JohnRTitor replied:
Commited. Resolving.

fufexan commented:
Commit names still aren't fixed.

JohnRTitor commented:
How do I amend previously created commits?

fufexan commented:
You can do `git rebase --interactive HEAD~4` to select the latest 4 commits (your commits), and change the `pick` keyword to `reword`. Then git will prompt you to edit each of the commits' messages.

JohnRTitor commented:
I have updated the commit titles. Please check.

fufexan commented:
The last two commits can be squashed into one. Using the same rebase mechanism, switch the last commit from `pick` to `squash`.

JohnRTitor commented:
Done. Thanks :)

fufexan reviewed (approved):
LGTM

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3716

JohnRTitor commented:
@NickCao  @SuperSandro2000 @wegank  gentle ping or if anyone willing to merge this

JohnRTitor commented:
@wegank I just rebased it :) no new changes

JohnRTitor commented:
@Mic92 could you please review it? Fufexan approved this change but still need a commiter to merge.

JohnRTitor commented:
> It worked fine before. Might this be caused by this PR or is this unrelated?

Working for me. Weird.
